File: 0047-permutations-ii/0047-permutations-ii.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        int n=nums.length;
        boolean visited[]=new boolean[n];
        List<List<Integer>>ans=new ArrayList<>();
        backtrack(ans,new ArrayList<>(),visited,nums);
        return ans;
        
    }
    public void backtrack( List<List<Integer>>ans,List<Integer>temp,boolean visited[],int nums[]){
        if(temp.size()==nums.length && !ans.contains(temp)){
            ans.add(new ArrayList<>(temp));
            return;
        }
        
        for(int i=0;i<nums.length;i++){
              if(visited[i]==true){
                continue;
              }
              visited[i]=true;
              temp.add(nums[i]);
              backtrack(ans,temp,visited,nums);
              visited[i]=false;
              temp.remove(temp.size()-1);
        }
    }
}

File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        ans =set()
        visited = [False] * len(nums)
        self.solve(nums, visited, [], ans)
        return [list(s) for s in ans]

    def solve(self, nums, visited, temp, ans):
        if len(temp) == len(nums):
            ans.add(tuple(temp.copy()))
            return
        
        for i in range(len(nums)):
            if visited[i]:
                continue
            visited[i] = True
            self.solve(nums, visited, temp + [nums[i]], ans)
            visited[i] = False
        